llvm/lib/SYCLLowerIR/ESIMD/LowerESIMD.cpp

```diff
@@ -1848,7 +1850,95 @@ void lowerGlobalsToVector(Module &M) {
+    if (F.hasFnAttribute(Attribute::AlwaysInline)) {
+      NeedInline = true;
+      continue;
+    }
+
+    // VC BE forbids 'alwaysinline' and "VCStackCall" on the same function.
+    // Such function may be used in other module, we cannot remove it
+    // after inlining.
+    if (F.hasFnAttribute(llvm::genx::VCFunctionMD::VCStackCall))
+      continue;
+
+    if (isESIMDKernel(F))
+      continue;
+
+    if (isSlmAllocatorConstructor(F) || isSlmAllocatorDestructor(F)) {
+      // slm_allocator constructor and destructor must be inlined
+      // to help SLM reservation analysis.
+      NeedInline |= markAlwaysInlined(F);
+      continue;
+    }
+
+    // TODO: The next code and comment was placed to ESIMDLoweringPass
+    // 2 years ago, when GPU VC BE did not support function calls and
+    // required everything to be inlined right into the kernel unless
+    // it had noinline or VCStackCall attrubute.
+    // This code migrated to here without changes, but... VC BE do support
+    //  the calls of spir_func these days, so this code may need re-visiting.
+    if (!F.hasFnAttribute(Attribute::NoInline))
+      NeedInline |= markAlwaysInlined(F);
+
+    if (!isSlmInit(F))
+      continue;
+
+    for (User *U : F.users()) {
+      auto *FCall = dyn_cast<CallBase>(U);
+      if (FCall && FCall->getCalledFunction() == &F) {
+        Function *GenF = FCall->getFunction();
+        // The original kernel (UserK) if often automatically separated into
+        // a spir_func (GenF) that is then called from spir_kernel (GenK).
+        // When that happens, the calls of slm_init<N>() originally placed
+        // in 'UserK' get moved to spir_func 'GenF', which creates wrong IR
+        // because slm_init() must be called only from a kernel.
+        // Fix it here: If 'GenF' has only 1 caller spir_kernel 'GenK',
+        // then inline 'GenF' to move slm_init call from spir_kernel 'GenK'.
+        SmallPtrSet<Function *, 1> GenFCallers;
+        for (User *GenFU : GenF->users()) {
+          auto *GenFCall = dyn_cast<CallBase>(GenFU);
+          if (GenFCall && GenFCall->getCalledFunction() == GenF) {
+            Function *GenK = GenFCall->getFunction();
+            GenFCallers.insert(GenK);
+          } else {
+            // Unexpected user of GenF. Do not require GenF inlining.
+            GenFCallers.clear();
+            break;
+          }
+        } // end for (User *GenFU : GenF->users())
+        if (GenFCallers.size() == 1 && isESIMDKernel(**GenFCallers.begin()))
+          NeedInline |= markAlwaysInlined(*GenF);
+      }
+    } // end for (User *U : F.users())
+  }
+  return NeedInline;
+}
+
+PreservedAnalyses SYCLLowerESIMDPass::run(Module &M,
+                                          ModuleAnalysisManager &MAM) {
+  // AlwaysInlinerPass is required for correctness.
+  bool ForceInline = prepareForAlwaysInliner(M);
+  if (ForceInline) {
+    ModulePassManager MPM;
+    MPM.addPass(AlwaysInlinerPass{});
+    MPM.run(M, MAM);
+  }
```

Comment:
I need careful review in this place because I am not sure if what I did is correct. It works, but might be wrong, e.g. ideologically wrong.

Reply:
IMO the ideal long term solution is either for VC to relax this requirement, or we ensure it in the front end.